inspector: add inspector.waitForConnection method

ak239 opened the pull request:
This method might be useful in case when we need to do something
with `inspector.url()` before waiting for connection, e.g. save it
to file.

With this method we can do something like:
```
inspector.open();
fs.writeFileSync(fileName, inspector.url());
inspector.waitForConnection();
```

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

ak239 commented:
@eugeneo please take a look.

ak239 commented:
Blocking here is exactly what I expect. My use case is:
1. call `inspector.open` to start inspector.
2. get `inspector.url()` and report it to protocol client.
3. wait until this client is connected to current Node.

It is emulation of `--inspect-brk` flag using inspector API. I think that as long as we have `--inspect-brk` flag, there is nothing bad in having the same feature as part of inspector API.

Alternative might be way to pass to node additional flag next to `--inspect-brk` that will force inspector to report `inspector.url()` in some way but it will be less flexible.

eugeneo commented:
Waiting happens before JS begins execution in the case of `--inspector-brk`...

Is it not possible to report URL before `inspector.open`? If the issue is with the GUID string generation, then it may be generated beforehand and returned by inspector API.

ak239 commented:
Is it possible to return full url including port and guid? It will work for my use case.

ak239 commented:
Unfortunately I start node with port 0 and let it detect port by itself. It sounds like I can not get full url until inspector.open is called, I think I have two options: consider adding another --inspector-something flag and if it is set - stored url somewhere, or merge this PR. Wdyt?

eugeneo commented:
Can't you just parse stderr since you seem to have access to the command line already (as you can pass the command line parameter)?

ak239 commented:
I pass arguments to child processes using NODE_OPTIONS, so parent process might mute or process stderr of child process by itself. At the same time I would like to be able to connect to any node process that started with environment that includes NODE_OPTIONS, in this case I don't have access to stderr stream.

eugeneo commented on doc/api/inspector.md:
```diff
@@ -49,6 +49,13 @@ and flow control has been passed to the debugger client.
 
+## inspector.waitForConnection()
+
+This method blocks until first client has connection, inspector should be open
```
My reading of this patch is that this will wait until:
1. A connection was made on a WS endpoint.
2. Runtime.runIfWaitingForDebugger command was issued

runIfWaitingForDebugger should be mentioned here to reduce confusion

eugeneo commented on src/inspector_js_api.cc:
```diff
@@ -259,6 +259,16 @@ void Open(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Agent* agent = env->inspector_agent();
+  if (!agent->IsActive()) {
+    env->ThrowError("inspector error, inspector.open should be called first");
```
Can this check be made from JavaScript code? I believe the guideline was not to throw exceptions from the native code.

eugeneo commented on src/inspector_agent.cc:
```diff
@@ -833,7 +833,8 @@ bool Agent::IsActive() {
 void Agent::WaitForConnect() {
   CHECK_NOT_NULL(client_);
-  client_->waitForFrontend();
+  if (!client_->hasConnectedSessions())
```
I am not sure this change is needed. If I understand correctly, the goal here is to wait for a specific frontend to connect. This check will subvert this expectation.

ak239 commented:
Unfortunately I lost original branch, so I open another PR and addressed all your comments there, please take a look: https://github.com/nodejs/node/pull/28453
